src/tokenserver/extractors.rs

```diff
@@ -0,0 +1,228 @@
+        })
+    }
+}
+
+/// Extracts data from the JWT in the Authorization header
+impl FromRequest for TokenserverRequest {
+    type Config = ();
+    type Error = Error;
+    type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;
+
+    fn from_request(req: &HttpRequest, payload: &mut Payload) -> Self::Future {
+        let req = req.clone();
+        let mut payload = payload.take();
+
+        Box::pin(async move {
+            let state = match req.app_data::<Data<ServerState>>() {
+                Some(s) => s,
+                None => {
+                    error!("⚠️ Could not load the app state");
+                    return Err(ValidationErrorKind::FromDetails(
+                        "Internal error".to_owned(),
+                        RequestErrorLocation::Unknown,
+                        Some("app_data".to_owned()),
+                        None,
+                    )
+                    .into());
+                }
+            };
+            let auth = BearerAuth::from_request(&req, &mut payload).await?;
+            let fxa_uid = {
+                let rsa_modulus = state.tokenserver_jwks_rsa_modulus.clone().ok_or_else(|| {
+                    error!("⚠️ Tokenserver JWK RSA modulus not set");
+                    ValidationErrorKind::FromDetails(
+                        "Internal error".to_owned(),
+                        RequestErrorLocation::Unknown,
+                        Some("app_data".to_owned()),
+                        None,
+                    )
+                })?;
+                let rsa_exponent =
+                    state.tokenserver_jwks_rsa_exponent.clone().ok_or_else(|| {
+                        error!("⚠️ Tokenserver JWK RSA exponent not set");
+                        ValidationErrorKind::FromDetails(
+                            "Internal error".to_owned(),
+                            RequestErrorLocation::Unknown,
+                            Some("app_data".to_owned()),
+                            None,
+                        )
+                    })?;
+                Self::get_fxa_uid(auth.token(), rsa_modulus, rsa_exponent)?
+            };
+
+            Ok(Self { fxa_uid })
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use actix_web::{http::Method, test::TestRequest, HttpResponse};
+    use jsonwebtoken::{encode, EncodingKey, Header};
+    use lazy_static::lazy_static;
+    use tokio::sync::RwLock;
+
+    use crate::db::mock::MockDbPool;
+    use crate::server::{metrics, ServerState};
+    use crate::settings::{Deadman, Secrets, ServerLimits, Settings};
+
+    use std::sync::Arc;
+    use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+    lazy_static! {
+        static ref SECRETS: Arc<Secrets> = Arc::new(Secrets::new("Ted Koppel is a robot").unwrap());
+        static ref SERVER_LIMITS: Arc<ServerLimits> = Arc::new(ServerLimits::default());
+    }
+
+    const RSA_PRIVATE_KEY: &str = "-----BEGIN RSA PRIVATE KEY-----\n\
```

Comment:
This is a key I generated specifically for the purposes of testing

Reply:
In this past, our security teams have advised me not to do even this. They suggested generating the keys at test time / somewhere in the repo, instead of checking them in.

Reply:
Ah, noted - thanks for the tip. I'll go ahead and do that then

Reply:
Oh, and to expand on my previous message: The reasoning was that if you see an RSA key in a repo, it should set off alarm bells. It's easier to just never set of the alarms, instead of continually noting that it's fine.

Reply:
Makes sense. And in this case, it literally set off alarm bells (in the form of a GitGuardian email)